python/mlc_llm/testing/debug_compare.py

```diff
@@ -0,0 +1,249 @@
+        debug_out : Path
+            the directory to dump the .npz files
+        """
+        _print_as_table(
+            sorted(
+                self.time_eval_results.items(),
+                key=lambda x: -(x[1][0] * x[1][1]),
+            )
+        )
+        self.time_eval_results = {}
+        self.visited = set([])
+        self.counter = 0
+
+    def skip_instrument(self, func, name, before_run, ret_val, *args):
+        if name.startswith("shape_func"):
+            return True
+        if self.counter < self.skip_rounds:
+            self.counter += 1
+            print(f"[{self.counter}] Skip validating {name}..")
+            return True
+        if name in self.visited:
+            if self.time_eval and name in self.time_eval_results:
+                record = self.time_eval_results[name]
+                self.time_eval_results[name] = (record[0], record[1] + 1)
+            return True
+        self.visited.add(name)
+        return False
+
+    def compare(
+        self,
+        name: str,
+        ref_args: List[tvm.nd.NDArray],
+        new_args: List[tvm.nd.NDArray],
+        ret_indices: List[int],
+    ):
+        super().compare(name, ref_args, new_args, ret_indices)
+
+        if self.time_eval and name not in self.time_eval_results:
+            res = self.mod.time_evaluator(
+                name, self.device, number=20, repeat=3  # , cache_flush_bytes=256 * 10**6
+            )(*new_args)
+            self.time_eval_results[name] = (res.mean, 1)
+            print(f"Time-eval result {name} on {self.device}:\n {res}")
+
+
+def get_instrument(args):
+    """Get the debug instrument from the CLI arguments"""
+    if args.cmp_device is None:
+        assert args.cmp_lib_path is None, "cmp_lib_path must be None if cmp_device is None"
+        args.cmp_device = args.device
+        args.cmp_lib_path = args.model_lib_path
+
+    if args.cmp_device == "iphone":
+        assert args.cmp_lib_path.endswith(".dylib"), "Require a dylib file for iPhone"
+        proxy_host = os.environ.get("TVM_RPC_PROXY_HOST", "127.0.0.1")
+        proxy_port = int(os.environ.get("TVM_RPC_PROXY_PORT", "9090"))
+        sess = rpc.connect(proxy_host, proxy_port, "iphone")
+        sess.upload(args.cmp_lib_path)
+        lib = sess.load_module(os.path.basename(args.cmp_lib_path))
+        cmp_device = sess.metal()
+    elif args.cmp_device == "android":
+        assert args.cmp_lib_path.endswith(".so"), "Require a so file for Android"
+        tracker_host = os.environ.get("TVM_TRACKER_HOST", "0.0.0.0")
+        tracker_port = int(os.environ.get("TVM_TRACKER_PORT", "9190"))
+        tracker = rpc.connect_tracker(tracker_host, tracker_port)
+        sess = tracker.request("android")
+        sess.upload(args.cmp_lib_path)
+        lib = sess.load_module(os.path.basename(args.cmp_lib_path))
+        cmp_device = sess.cl(0)
+    else:
+        lib = tvm.runtime.load_module(
+            os.path.join(
+                args.artifact_path,
```

Comment:
I'm a little confused where this `artifact_path` comes from. Is this supposed to be the `model` folder? Or shall we just use the `cmp_lib_path` instead?

Reply:
it was historically the path of the dist, ia gree having cmp-lib-path now make smore sense